src/object/srv_obj_migrate.c

```diff
@@ -3271,8 +3272,70 @@ migrate_obj_ult(void *data)
 		}
 	}
 
+	if (!tls->mpt_reintegrating_check) {
+		struct pool_target *target;
+		struct ds_pool *pool = tls->mpt_pool->spc_pool;
+
+		ABT_rwlock_rdlock(pool->sp_lock);
+		rc = pool_map_find_target_by_rank_idx(pool->sp_map, dss_self_rank(),
+						      dss_get_module_info()->dmi_tgt_id,
+						      &target);
+		ABT_rwlock_unlock(pool->sp_lock);
+		D_ASSERT(rc == 1);
+
+		if (target->ta_comp.co_status == PO_COMP_ST_UP &&
+		    target->ta_comp.co_in_ver <= tls->mpt_version)
+			tls->mpt_reintegrating = 1;
+		D_DEBUG(DB_REBUILD, "status %u in version %u, mpt_reintegrating %d\n",
+			target->ta_comp.co_status, target->ta_comp.co_in_ver,
+			tls->mpt_reintegrating);
+
+		tls->mpt_reintegrating_check = 1;
+	}
+
+	if (tls->mpt_reintegrating) {
+		struct ds_cont_child *cont_child = NULL;
+
+		/* check again to see if the container is being destroyed. */
+		migrate_get_cont_child(tls, arg->cont_uuid, &cont_child, false);
+		if (cont_child != NULL && !cont_child->sc_stopping) {
+			if (vos_oi_exist(cont_child->sc_hdl, arg->oid)) {
+				rc = vos_dtx_cmt_reindex(cont_child->sc_hdl);
+				if (rc < 0) {
+					ds_cont_child_put(cont_child);
+					D_GOTO(out, rc);
+				}
+				stable_epoch = vos_cont_get_global_stable_epoch(cont_child->sc_hdl);
+			} else {
+				/* If the object does not exist on the local target at all,
+				 * it is either created after stable epoch or migrated to this
+				 * target due to co-location. let's set stable epoch to 0,
+				 * migrate the whole object.
+				 */
+				stable_epoch = 0;
+			}
+		}
+
+		if (cont_child)
+			ds_cont_child_put(cont_child);
+	}
+
 	for (i = 0; i < arg->snap_cnt; i++) {
-		epr.epr_lo = i > 0 ? arg->snaps[i - 1] + 1 : 0;
+		daos_epoch_t lower_epoch = 0;
+
+		if (arg->snaps[i] < stable_epoch) {
+			D_DEBUG(DB_REBUILD, DF_CONT " obj " DF_UOID " skip snap "DF_X64
+				" < stable " DF_X64 "\n", DP_CONT(arg->pool_uuid, arg->cont_uuid),
+				DP_UOID(arg->oid), arg->snaps[i], stable_epoch);
+			continue;
+		} else {
```

Comment:
unnecessary `else` since above `continue`.

Reply:
right, it can be removed.